picoquic/bytestream.c

```diff
@@ -0,0 +1,306 @@
+/*
+* Author: Christian Huitema
+* Copyright (c) 2018, Private Octopus, Inc.
+* All rights reserved.
+*
+* Permission to use, copy, modify, and distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL Private Octopus, Inc. BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "bytestream.h"
+#include "picoquic_internal.h"
+
+static int bytestream_error(bytestream * s);
+
+bytestream * bytereader_init(bytestream * s, const void * bytes, size_t nb_bytes)
+{
+    s->data = (uint8_t*)bytes;
+    s->size = nb_bytes;
+    s->ptr = 0;
+
+    return s;
+}
+
+bytestream * bytewriter_init(bytestream_buf * s)
+{
+    s->s.data = s->buf;
+    s->s.size = BYTESTREAM_MAX_BUFFER_SIZE;
+    s->s.ptr = 0;
+
+    return &s->s;
+}
+
```

Comment:
Why can we specify a size in the "reader_init" command but not in the "writer_init" equivalent?

Reply:
Good point.

The naming is not the best here. It depends more on how you want to use the bytestream. If you already have memory allocated, use the reader init. If not, like in a usual write use case, use the writer init, which provides temporary memory of 2048 bytes, which should be enough to write whole packets. If you need bigger buffers, use the bytestream_alloc initialization, which allocates the memory on the heap.

I'll rework the naming.

Reply:
I've also added a size parameter to the "writer_init". That totally makes sense! Thanks for pointing that out 👍